sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubAvroIT.java

```diff
@@ -0,0 +1,95 @@
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.pubsub;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasProperty;
+
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.utils.AvroUtils;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.hamcrest.Matcher;
+import org.joda.time.Instant;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Integration tests for querying Pubsub AVRO messages with SQL. */
+@RunWith(JUnit4.class)
+public class PubsubAvroIT extends PubsubTableProviderIT {
+  private static final Schema NAME_HEIGHT_KNOWS_JS_SCHEMA =
+      Schema.builder()
+          .addNullableField("name", Schema.FieldType.STRING)
+          .addNullableField("height", Schema.FieldType.INT32)
+          .addNullableField("knowsJavascript", Schema.FieldType.BOOLEAN)
+          .build();
+
+  private static final Schema NAME_HEIGHT_SCHEMA =
+      Schema.builder()
+          .addNullableField("name", Schema.FieldType.STRING)
+          .addNullableField("height", Schema.FieldType.INT32)
+          .build();
+
+  @Override
+  protected String getPayloadFormat() {
+    return "avro";
+  }
+
+  @Override
+  protected PCollection<String> applyRowsToStrings(PCollection<Row> rows) {
+    return rows.apply(
+        MapElements.into(TypeDescriptors.strings())
+            .via(
+                (Row row) ->
+                    new String(
+                        AvroUtils.getRowToAvroBytesFunction(row.getSchema()).apply(row), UTF_8)));
+  }
+
+  @Override
+  protected PubsubMessage messageIdName(Instant timestamp, int id, String name) {
+    Row row = row(PAYLOAD_SCHEMA, id, name);
+    return message(timestamp, AvroUtils.getRowToAvroBytesFunction(PAYLOAD_SCHEMA).apply(row));
+  }
+
+  @Override
+  protected Matcher<PubsubMessage> matcherNames(String name) {
+    Schema schema = Schema.builder().addStringField("name").build();
+    Row row = row(schema, name);
+    return hasProperty("payload", equalTo(AvroUtils.getRowToAvroBytesFunction(schema).apply(row)));
+  }
+
+  @Override
+  protected Matcher<PubsubMessage> matcherNameHeight(String name, int height) {
+    Row row = row(NAME_HEIGHT_SCHEMA, name, height);
+    return hasProperty(
+        "payload", equalTo(AvroUtils.getRowToAvroBytesFunction(NAME_HEIGHT_SCHEMA).apply(row)));
+  }
+
+  @Override
+  protected Matcher<PubsubMessage> matcherNameHeightKnowsJS(
+      String name, int height, boolean knowsJS) {
+    Row row = row(NAME_HEIGHT_KNOWS_JS_SCHEMA, name, height, knowsJS);
+    return hasProperty(
+        "payload",
+        equalTo(AvroUtils.getRowToAvroBytesFunction(NAME_HEIGHT_KNOWS_JS_SCHEMA).apply(row)));
```

Comment:
These functions should directly verify the Avro data rather than converting to a Row if possible

Reply:
Done.